lib/server/routes/buckets.js

```diff
@@ -1413,6 +1416,332 @@ BucketsRouter.prototype.getStorageUsage = function (req, res) {
+
+            createdShards.push({
+              index: uploadObject.index,
+              size: uploadObject.data_size,
+              _id: savedShard._id,
+            });
+
+            async.eachLimit(
+              contracts,
+              1,
+              (contract, cb) => {
+                const mirrorToCreate = new Mirror({
+                  isEstablished: true,
+                  shardHash: hash,
+                  contact: contract.nodeID,
+                  contract,
+                  token: '',
+                  created: new Date()
+                });
+
+                mirrorToCreate.save(cb);
+              },
+              cb
+            );
+          });
+        });
+      },
+      (err) => {
+        if (err) {
+          return next(err);
+        }
+
+        const bucketEntryToCreate = new BucketEntry({
+          bucket: idBucket,
+          index,
+          created: new Date(),
+          shards: createdShards,
+        });
+
+        bucketEntryToCreate.save((err, savedBucketEntry) => {
+          if (err) {
+            return next(err);
+          }
+
+          const savedBucketEntryObject = savedBucketEntry.toObject();
+          res.status(200).send(savedBucketEntryObject);
+        });
+      }
+    );
+  });
+};
+
+BucketsRouter.prototype.getDownloadLinks = function (req, res, next) {
+  const { id:bucketId, file:fileId } = req.params;
+  const { BucketEntry, Shard, Mirror, Contact } = this.storage.models;
+
+  if (!bucketId) {
+    return next(new errors.BadRequestError('No bucket id'));
+  }
+
+  if (!fileId) {
+    return next(new errors.BadRequestError('No file id'));
+  }
+
+  BucketEntry.findOne({
+    bucket: bucketId,
+    _id: fileId
+  }, (err, entry) => {
+    if (err) {
+      return next(new errors.InternalError(err.message));
+    }
+
+    if (!entry) {
+      return next(new errors.NotFoundError('File not found'));
+    }
+
+    const { shards } = entry;
+    async.mapLimit(
+      shards,
+      1,
+      (shardId, nextShardId) => {
+
+        Shard.findOne({
+          _id: shardId
+        }, (err, shard) => {
+          if (err) {
+            return nextShardId(new errors.InternalError(err.message));
+          }
+
+          if (!shard) {
+            return nextShardId(new errors.NotFoundError('Shard not found'));
+          }
+
+          const { hash } = shard;
+
+          Mirror.findOne({
+            shardHash: hash
+          }, (err, mirror) => {
+            if (err) {
+              return nextShardId(err);
+            }
+
+            Contact.findOne({
+              _id: mirror.contact
+            }, (err, contact) => {
+              if (err) {
+                return nextShardId(err);
+              }
+
+              if (!contact) {
+                return nextShardId(new errors.NotFoundError('Contact not found'));
+              }
+
+              const { address, port } = contact;
+              const farmerUrl = `http://${address}:${port}/download/link/${hash}`;
```

Comment:
shouldn't this be the uuid?

Reply:
yes, should be

Reply:
ping me when it is changed to proceed with this